M2/Macaulay2/packages/MacaulayPosets.m2

```diff
@@ -0,0 +1,1956 @@
+    
+    instance(scan((vertices source f), i->if not (f(i)==g(i)) then break false), Nothing)
+);
+PosetMap*PosetMap := PosetMap => (f,g) -> (
+    map(target f, source g, apply(vertices source g, p -> f(g(p))))
+);
+image(PosetMap, List) := List => (f, L) -> (
+    unique apply(select(L, l->member(l,vertices source f)), p->f(p))
+);
+image(PosetMap) := List => (f) -> (
+    image(f, vertices source f)
+);
+isInjective(PosetMap) := Boolean => (f) -> (
+    --use finiteness
+    #(image f) == #(vertices source f)
+); 
+isSurjective(PosetMap) := Boolean => (f) -> (
+    isSubset(vertices target f, image f)
+);
+isRankPreserving = method();
+isRankPreserving(PosetMap) := Boolean => (f) -> (
+    if (not isRanked source f) or (not isRanked target f) then return false;
+    
+    targetRankFunction := rankFunction target f;
+    targetRank := new HashTable from apply(#(vertices target f), i -> (vertices target f)#i => targetRankFunction#i);
+    
+    sourceRankFunction := rankFunction source f;
+    sourceRank := new HashTable from apply(#(vertices source f), i -> (vertices source f)#i => sourceRankFunction#i);
+    
+    instance(scan(vertices source f, p -> (
+        if not (sourceRank#p == targetRank#(f(p))) then break false;
+    )), Nothing)
+);
+
+--------------------------------------------------
+--- Getting posets of polynomial rings with any homogeneous ideal
+--------------------------------------------------
+
+-- Suppose s is a monomial in a ring S. This returns what degree s would be if S was replaced by newRing(S, Degrees=>toList((numgens S):1))
+theDegree = (s) -> (
+    S := ring s;
+    
+    if s == 0_S then return -infinity;
+    
+    sum := 0;
+    for g in (gens S) do (
+        if not (g == 0_S) then (
+            sum = sum + degree(g,s);
+        );
+    );
+    sum
+)
+
+isHomogeneousWrtVars = (I) -> (
+    S := (ring I)/I;
+    isHomogeneous newRing(S, Degrees=>toList((numgens S):1))
+);
+
+-- First function is to get all the unique monomial representations in the quotient ring. The function super basis is the cornerstone for this. Everything else is just manipulating the data. 
+
+getMons = method(TypicalValue => List, Options => { MaxDegree => 10 } );
+
+getMons(PolynomialRing, Ideal) := options -> (R, I) -> (
+    M := {};
+    if not isField baseRing R then return "The base ring of the polynomial ring must be a field.";
+    if not isHomogeneousWrtVars I then return "The given ring must be homogeneous.";
+    try R/I then S := R/I else return "The first input needs to be a polynomial ring R, second input needs to be an ideal of R";
+    
+    try super basis(S) then ( -- If the quotient ring has a finite number of elements (ie. the super basis returns a finite number of monomials): 
+        M = super basis(S);
+        M = flatten entries M;
+    ) else ( -- If the quotient ring doesn't have a finite number of elements:
+        M = new MutableList;
```

Comment:
Do you know what the length of the list will be ahead of time?  The way this is currently implemented (calling `M#(#M)` a bunch), we'll be re-sizing `M` and creating a brand new list every time.

Reply:
And if you don't know the length, then `while ... list` can be useful.  It expands the list as needed, but not at every step.  (I think it does something like double the size when it's about to run out of space.)